Bugfix: Cert-csi taking incorrect size for creating PVC

mgandharva opened the pull request:
# Description
Updated the size of the new PVC to match the snapshot size, (which was initially created from the PVC, instead of using the original PVC size, because the PowerMax array/storage rounds the PVC size to the next cylinder boundary, resulting in a slightly larger PVC and snapshot size in the array). 

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| https://github.com/dell/csm/issues/1792|

# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have maintained backward compatibility

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [x] Replication Suite ran successfully.

![image](https://github.com/user-attachments/assets/7b4373bc-ffd8-4ed3-9d7e-5a456f6447f2)


- [x] Snapshot Suite ran successfully.

![image](https://github.com/user-attachments/assets/d2943263-aaa9-4389-aa8f-6a330c0cd8cf)

- [x] Capacity-tracking Suite ran successfully.

![image](https://github.com/user-attachments/assets/f8fa9135-4db8-4e88-9a38-72185d9f6d91)

chimanjain commented on pkg/testcore/suites/perf-suites.go:
```diff
@@ -1523,6 +1527,31 @@ func (ss *SnapSuite) Run(ctx context.Context, storageClass string, clients *k8sc
+	}
+	log.Debugf("kubectl command output: %s", output)
+	// Parse the JSON output using jq
+	cmd = exec.Command("jq", "-r", ".status.restoreSize")
```
Is it possible to remove `jq` dependency?

mgandharva replied:
Code updated, Please re-review.

chimanjain replied:
Looks good

kumarp20 commented:
@mgandharva @shanmydell @sharmilarama Is this the correct fix that we want to put in ? I would assume that it is normal for a user to expect that create snap from a 2 GB volume and use that snap to create another 2 GB volume would succeed. In our case it doesnt since CSM 1.12 , cert-csi tests show that and we are fixing the cert-csi tests to create the new volume with a bigger size. Shouldnt our fix be to set the restoreSize of the snapshot such that it can be restored to a 2 GB volume ?

mgandharva commented:
> @mgandharva @shanmydell @sharmilarama Is this the correct fix that we want to put in ? I would assume that it is normal for a user to expect that create snap from a 2 GB volume and use that snap to create another 2 GB volume would succeed. In our case it doesnt since CSM 1.12 , cert-csi tests show that and we are fixing the cert-csi tests to create the new volume with a bigger size. Shouldnt our fix be to set the restoreSize of the snapshot such that it can be restored to a 2 GB volume ?

@kumarp20  
"When we create a PVC with a size of 3GB in our template, the PowerMax array/storage rounds this size to the next cylinder boundary, resulting in an actual size of 3.X GB. Consequently, the snapshot created from this PVC will be larger than the original 3GB.

When we attempt to create a new volume from this snapshot using the original template size of 3GB, it results in an error because the target volume size is smaller than the snapshot size. Therefore, the correct approach is to create the new volume with a size equal to the snapshot size.

This can be achieved in two ways:

1. Retrieving the snapshot size from the Kubernetes cluster (which I have implemented).
2. Querying the array using the API to get the exact size.

By ensuring the new volume matches the snapshot size, we eliminate the errors and align with best practices.

jooseppi-luna commented:
Two comments -- 
1. Agreed with @chimanjain -- let's remove the jq dependency, we shouldn't have dependencies on outside tools. Also, is it possible to programmatically read the snapshot size using e.g. client-go or the k8s clients we have implemented in cert-csi? That would be a more robust solution.
2. Can we fix the description of the PR -- the snapshot size is not greater than the requested size of the PVC because of metadata. It's because the original PVC is greater in size than what was requested, because PowerMax always rounds the requested size to the next cylinder boundary.

mgandharva commented:
> Two comments --
> 
> 1. Agreed with @chimanjain -- let's remove the jq dependency, we shouldn't have dependencies on outside tools. Also, is it possible to programmatically read the snapshot size using e.g. client-go or the k8s clients we have implemented in cert-csi? That would be a more robust solution.
> 2. Can we fix the description of the PR -- the snapshot size is not greater than the requested size of the PVC because of metadata. It's because the original PVC is greater in size than what was requested, because PowerMax always rounds the requested size to the next cylinder boundary.

Updated, please re-review.

chimanjain reviewed (dismissed):
Looks good from my side

kumarp20 commented:
> > @mgandharva @shanmydell @sharmilarama Is this the correct fix that we want to put in ? I would assume that it is normal for a user to expect that create snap from a 2 GB volume and use that snap to create another 2 GB volume would succeed. In our case it doesnt since CSM 1.12 , cert-csi tests show that and we are fixing the cert-csi tests to create the new volume with a bigger size. Shouldnt our fix be to set the restoreSize of the snapshot such that it can be restored to a 2 GB volume ?
> 
> @kumarp20 "When we create a PVC with a size of 3GB in our template, the PowerMax array/storage rounds this size to the next cylinder boundary, resulting in an actual size of 3.X GB. Consequently, the snapshot created from this PVC will be larger than the original 3GB.
> 
> When we attempt to create a new volume from this snapshot using the original template size of 3GB, it results in an error because the target volume size is smaller than the snapshot size. Therefore, the correct approach is to create the new volume with a size equal to the snapshot size.
> 
> This can be achieved in two ways:
> 
> 1. Retrieving the snapshot size from the Kubernetes cluster (which I have implemented).
> 2. Querying the array using the API to get the exact size.
> 
> By ensuring the new volume matches the snapshot size, we eliminate the errors and align with best practices.

@shanmydell @sharmilarama @anandrajak1 Product Owners, Please confirm if this this is how the POs expect this defect to be fixed. Note that it does change the user experience since before CSM 1.12, it was possible to snap a 3GB volume and use the snap to create another 3GB volume and that is not possible any more with a change that went in CSM 1.12. This PR is updating the tests that now do not pass, to pass. Hence would like the POs to confirm this is the behavior they want users to experience.

shanmydell commented:
@mgandharva : @kumarp20 is correct here, taking a snapshot and restoring are general use case for PMax customer, we need to identify the root cause and fix it at the driver

mgandharva commented:
Since the fix has been implemented in the csi-powermax repository (see PR: https://github.com/dell/csi-powermax/pull/491), this pull request is no longer necessary.
